src/pubsub/publish_sink.rs

```diff
@@ -388,11 +421,24 @@ where
                 // original to reuse for retries/errors/responses. On the bright side, message
                 // payloads are Bytes which are essentially Arc'ed, so alloc clones mostly apply to
                 // the attribute maps
-                break match client.publish(request.clone()).await {
-                    Ok(response) => Ok(response.into_inner()),
+                let mut tonic_request = tonic::Request::new(request.clone());
+                // set the grpc-timeout header to inform the server of our deadline
+                tonic_request.set_timeout(timeout);
+
+                let publish = client.publish(tonic_request);
+
+                // apply a timeout to the publish
+                let publish_fut = tokio::time::timeout(timeout, publish).map_err(
```

Comment:
Yes. Client side is the more obvious one, in case the server never responds, or a network problem, or whatever. Server side is more of a good convention; if the server takes too long and knows the client will timeout, the server is allowed to free resources after that time, which can help responsiveness overall when under load.